ui/src/ducks/login.js

```diff
@@ -28,23 +36,19 @@ export default function reducer(state = defaultState, action = {}) {
 }
 
 // Action Creators
-export const authenticateSaltApiAction = payload => {
-  return { type: AUTHENTICATE_SALT_API, payload };
-};
-
-export const setSaltAuthenticationSuccessAction = payload => {
+export const setSaltAuthenticationSuccessAction = (payload: ApiSalt.SaltToken) => {
   return {
     type: SALT_AUTHENTICATION_SUCCESS,
     payload,
   };
 };
 
 // Sagas
-export function* authenticateSaltApi() {
-  const api = yield select(apiConfigSelector);
-  const user = yield select(state => state.oidc.user);
-  const result = yield call(ApiSalt.authenticate, user);
-  if (!result.error) {
+export function* authenticateSaltApi(): Generator<any, void, any> {
+  const api: Config | null = yield select(apiConfigSelector);
```

Comment:
(ugh, even if flow was able to infer the return type of a selector, it can't guess the value injected by the coroutine's "owner"... some big limitation of sagas here)

Reply:
Yes the simplest way I found to type complex saga with different return type over iterations is to type `Next` as any and  force the type of each of the variables 😞 . Even through this is not directly linked to redux but more to typing, IMO redux and redux sage are adding a lot of complexity for such simple tasks as data fetching, I prefer using libraries such as  `react-query` or `swr` that greatly ease data fetching and caching use cases and that are way more easier to type as in most cases the typing would be inferred from API clients typing.